akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery.compress
+
+import scala.collection.immutable
+
+// TODO describe / think about how we could handle an actor dying and removing from "top"
+// TODO this is the scenario when "super traffic actor" goes away and we want to not use up the slot in HH
+
+/**
+ * INTERNAL API
+ *
+ * Mutable.
+ *
+ * Keeps a number of specific heavy hitters around in memory.
+ *
+ * See also Section 5.2 of http://dimacs.rutgers.edu/~graham/pubs/papers/cm-full.pdf
+ * for a discussion about the assumptions made and guarantees about the Heavy Hitters made in this model.
+ * We assume the Cash Register model in which there are only additions, which simplifies HH detecion significantly.
+ */
+private[remote] sealed class HeavyHitters[T](strategy: HeavyHittersStrategy[T]) {
+  // TODO technically could be optimised depending on strategy (priority queue?) (store in order of counts, drop last if exceeded size)
+  private[this] var _items = List.empty[HeavyHitter[T]]
+
+  def items: immutable.Set[T] = _items.map(_.item)(scala.collection.breakOut).toSet // TODO avoid the toSet? does anyone care?
+
+  /**
+   * Attempt adding item to heavy hitters set, if it does not fit in the top yet,
+   * it will be dropped and the method will return `false`.
+   *
+   * @return `true` if the added item is indeed a heavy hitter
+   */
+  def update(item: T, count: Long): Boolean =
+    _items.find(_.item == item) match { // TODO optimise? could it be the same case always?
```

Comment:
if using a PriorityQueue, you need a linear traversal here, but since it is Array backed it should be cache friendly. We can do even better with a custom queue if:
 - maintain a queue built from two separate arrays, one containing item hash, the other the actual item themself
 - search in the first queue by hash value (no pointer chasing since it is an Array[Long]) and check the second queue for equality if a matching hash is found
 - of course the two queues must be kept in sync, I.e. it is *one* priority queue just blasted into multiple arrays for cache friendliness (since we have no record types)